src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/ChangelogMigration.kt

```diff
@@ -0,0 +1,130 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.config
+
+import mu.KLogging
+import org.springframework.context.annotation.Configuration
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.dto.ReferralAmendmentDetails
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Changelog
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ReferralDetails
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.AuthUserRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ChangelogRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ReferralDetailsRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.AmendTopic
+import java.time.OffsetDateTime
+import java.time.temporal.ChronoUnit
+import java.util.UUID
+import javax.annotation.PostConstruct
+
+@Configuration
+class ChangelogMigration(
+  private val referralDetailsRepository: ReferralDetailsRepository,
+  private val changelogRepository: ChangelogRepository,
+  private val authUserRepository: AuthUserRepository
+) {
+  companion object : KLogging()
+
+  @PostConstruct
+  fun migrateDataFromReferralDetailsToChangeLog() {
+
+    logger.info("migration of referral details to change log starts")
+
+    // Get all the referral details where the superseded id not null
+
+    val referralDetails = referralDetailsRepository.findAll()
+
+    referralDetails.filter { it.supersededById != null }.forEach {
+      try {
+        val changeLogs = changelogRepository.findByReferralIdOrderByChangedAtDesc(it.referralId)
+        // Do this migration when there is no change log data
+        if (changeLogs.isEmpty()) {
+          migrateReferralToChangelogTable(it)
```

Comment:
_concurrency_

Possibility of stale data leading to multiple appends:

```kotlin
// first transaction
val changeLogs = changelogRepository.findByReferralIdOrderByChangedAtDesc(it.referralId)

// decision based on in-memory code
if (changeLogs.isEmpty()) {

  // second transaction
  migrateReferralToChangelogTable(it)
}
```

If 
- this code is running concurrently;
- someone amends referrals business as usual,

then the decision of emptiness might be stale already, and we would execute append on an already non-empty state.

Introducing a transaction boundary for `referral_detail` and `changelog` scoped to one specific referral ID would prevent concurrent updates until the check-and-write pattern finishes.